src/main/java/com/google/research/bleth/servlets/EnqueueExperimentServlet.java

```diff
@@ -0,0 +1,234 @@
+            values.add(PropertyWrapper.create(parameter, value));
+        }
+        return ImmutableSet.copyOf(values);
+    }
+
+    private Set<PropertyWrapper> createDoubleValues(HttpServletRequest request, String parameter) {
+        double lower = Double.parseDouble(request.getParameter("lower" + upperCaseFirstChar(parameter)));
+        double upper = Double.parseDouble(request.getParameter("upper" + upperCaseFirstChar(parameter)));
+        double step = Double.parseDouble(request.getParameter("step" + upperCaseFirstChar(parameter)));
+        Set<PropertyWrapper> values = new HashSet<>();
+        for (double value = lower; value <= upper; value += step) {
+            values.add(PropertyWrapper.create(parameter, value));
+        }
+        return ImmutableSet.copyOf(values);
+    }
+
+    private String upperCaseFirstChar(String s) {
+        return s.substring(0, 1).toUpperCase() + s.substring(1);
+    }
+
+    private AppEngineHttpRequest toHttpRequest(List<PropertyWrapper> configuration,
+                                               int beaconsNum, Key experimentId, String experimentTitle) throws UnsupportedEncodingException {
+        Map<String, String> bodyMap = new HashMap<>();
+        bodyMap.put(Schema.SimulationMetadata.description,
+                "Simulation attached to experiment: " + experimentTitle);
+        bodyMap.put(Schema.SimulationMetadata.beaconsNum, String.valueOf(beaconsNum));
+        bodyMap.put(Schema.SimulationMetadata.beaconMovementStrategy, defaultMovementStrategy.toString());
+        bodyMap.put(Schema.SimulationMetadata.observerMovementStrategy, defaultMovementStrategy.toString());
+        bodyMap.put(Schema.SimulationMetadata.observerAwakenessStrategy, defaultAwakenessStrategy.toString());
+        bodyMap.put(Schema.ExperimentsToSimulations.experimentId, KeyFactory.keyToString(experimentId));
+        configuration.forEach((propertyWrapper -> {
+            bodyMap.put(propertyWrapper.property(), String.valueOf(propertyWrapper.value()));
+        }));
+
+        return AppEngineHttpRequest.newBuilder()
+                .setRelativeUri("/new-experiment-simulation")
+                .setHttpMethod(HttpMethod.POST)
+                .putHeaders("Content-Type", "application/x-www-form-urlencoded;charset=UTF-8;")
+                .setBody(ByteString.copyFromUtf8(toQueryString(bodyMap)))
+                .build();
+    }
+
+    private String toQueryString(Map<String, String> params) throws UnsupportedEncodingException {
+        List<String> keyValuePairs = new ArrayList<>();
+        for (String param : params.keySet()) {
+            keyValuePairs.add(URLEncoder.encode(param, StandardCharsets.UTF_8.toString()) + "=" +
+                    URLEncoder.encode(params.get(param), StandardCharsets.UTF_8.toString()));
+        }
+        return String.join("&", keyValuePairs);
+    }
+
+    private boolean validateArguments(List<PropertyWrapper> configuration) {
```

Comment:
Don't you already have a similar validation for simulation creation? (How _do_ you validate user-submitted simulations?)

Reply:
The SimulationBuilder does have such method:

```java
void validateArguments() {
            checkArgument(rowNum > 0 && colNum > 0,
                    "Board dimensions must be positive.");
            checkArgument(beaconsNum > 0 && observersNum > 0,
                    "Number of beacons and number of observers must be positive.");
            checkArgument(awakenessCycle > 0 && awakenessDuration > 0,
                    "Both awakeness cycle and duration must be positive.");
            checkArgument(awakenessCycle >= awakenessDuration,
                    "Awakeness cycle must be greater or equal to duration.");
            checkArgument(transmissionThresholdRadius > 0,
                    "Transmission threshold radius must be positive.");
            checkArgument(maxNumberOfRounds > 0,
                    "Maximum number of rounds must be positive.");
            checkNotNull(beaconMovementStrategyType, "No beacon movement strategy has been set.");
            checkNotNull(observerMovementStrategyType, "No observer movement strategy has been set.");
            checkNotNull(awakenessStrategyType, "No awakeness strategy type has been set.");
        }
```

This method accesses the builder's fields so it cannot be used outside the class. 
I thought of implementing a utility validation function that will be used by both the builder and the servlet (and gets the parameters as a map, for example) - but then it became too messy: the builder and the servlet need to convert their fields/configuration into a map (or any other DS the utility function accepts as an argument) each time they want to use this utility method.

On the other hand I do understand why this is better to have both the builder and the servlet validation depend on the same generic method. I will modify it this way (unless you believe the potential code-complexity doesn't worth it).

@nolaviz

Reply:
Can you use a `SimulationBuilder` here, instead of 12 separate primitive variables?

Reply:
How about something like:

```java
private boolean validateArguments(List<PropertyWrapper> configuration) {
        Map<String, Number> properties = configuration.stream()
                .collect(Collectors.toMap(PropertyWrapper::property, PropertyWrapper::value));

        try {
            new TracingSimulation.Builder()
                    .setMaxNumberOfRounds(properties.get(Schema.SimulationMetadata.roundsNum).intValue())
                    .setRowNum(properties.get(Schema.SimulationMetadata.rowsNum).intValue())
                    .setColNum(properties.get(Schema.SimulationMetadata.colsNum).intValue())
                    .setBeaconsNum(properties.get(Schema.SimulationMetadata.beaconsNum).intValue())
                    .setObserversNum(properties.get(Schema.SimulationMetadata.observersNum).intValue())
                    .setBeaconMovementStrategyType(defaultMovementStrategy)
                    .setObserverMovementStrategyType(defaultMovementStrategy)
                    .setAwakenessStrategyType(defaultAwakenessStrategy)
                    .setAwakenessCycle(properties.get(Schema.SimulationMetadata.awakenessCycle).intValue())
                    .setAwakenessDuration(properties.get(Schema.SimulationMetadata.awakenessDuration).intValue())
                    .setTransmissionThresholdRadius(properties.get(Schema.SimulationMetadata.transmissionThresholdRadius)
                            .doubleValue())
                    .build();
        } catch (Exception e) {
            return false;
        }
        return true;
    }
```

@nolaviz

Reply:
This works, yeah, but it's a bit ugly...
Consider leaving a TODO in place to refactor `TracingSimulation.Builder.validateArguments()` so it's usable in more contexts (but we can leave actual refactoring for later, for now).